src/main/java/fi/aalto/cs/apluscourses/intellij/model/IntelliJModule.java

```diff
@@ -77,8 +69,10 @@ protected int resolveStateInternal() {
 
   @Override
   public void load() throws ComponentLoadException {
+    ModuleManager moduleManager = project.getModuleManager();
```

Comment:
This is actually one of those places I mentioned where I'm uncertain if a read action is needed or not

Reply:
You mean just accessing ModuleManager because the actual thing is done inside a write action? Surely, it's possible that getting ModuleManager instace w/o read lock causes errors. In this case, it could be easily fixed by moving call to `getModuleManager()` inside the write action.

What I think is more concerning case is `resolveStateInternal()` method above because it's called multiple times, and making it read action would be unfortunate.

Reply:
Yep, I'm referring specifically to the `ModuleMaager` access here.
> You must not access the model outside a read or write action. The corresponding objects are not guaranteed to survive between several consecutive read actions. As a rule of thumb, whenever starting a read action, check if PSI/VFS/project/module is still valid.

 I guess that means to be safe the module manager access should be moved inside the write action. It would be neater if the read action was done inside `APlusProject` so that clients don't have to worry about wrapping `APlusProject` method calls inside read actions, but then we couldn't conveniently put it inside the write action here. According to the quote we should also check here that the project is still valid, but I'm not 100% sure what that would mean in practice.

It would indeed be unfortunate if `resolveStateInternal` needs a read action. One option is to fine-tune the read actions inside `APlusProject`, since some things such as checking the existence of a directory don't need a read action. 

If I understand [this](https://www.jetbrains.org/intellij/sdk/docs/basics/architectural_overview/general_threading_rules.html) correctly, the real problem is long read actions, so perhaps multiple short consecutive read actions aren't too bad, since the UI thread can get write actions in between, but we should consider this carefully.

Reply:
I set those accesses that are inside `fi.aalto.cs.apluscourses.intellij.model` to use always read and write actions. I didn't notice any issues but let's see.